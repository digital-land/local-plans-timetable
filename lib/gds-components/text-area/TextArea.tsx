/**
 * GDS component: https://design-system.service.gov.uk/components/textarea/
 */
import cn from "classnames";
import { CharacterCount } from "../character-count/CharacterCount";

interface TextAreaProps {
  label?: string;
  hint?: string;
  onChange: (value: string) => void;
  value: string;
  id: string;
  characterLimit?: number;
  error?: string;
}

export const TextArea = ({
  label,
  hint,
  onChange,
  value,
  characterLimit,
  error,
  id,
}: TextAreaProps) => {
  return (
    <div
      className={cn("govuk-form-group govuk-!-width-two-thirds", {
        "govuk-form-group--error": error,
      })}
    >
      {label && (
        <label className="govuk-label govuk-label--m" htmlFor={id}>
          {label}
        </label>
      )}
      {hint && (
        <div id={`${id}-hint`} className="govuk-hint">
          {hint}
        </div>
      )}
      {error && (
        <p className="govuk-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </p>
      )}
      <textarea
        className="govuk-textarea govuk-!-margin-bottom-2"
        data-testid={`${label ? label.replace(/ /g, "-") : "input"}-text-area`}
        rows={5}
        value={value}
        onChange={(e) => onChange(e.target.value)}
        id={id}
        aria-describedby={
          cn({
            [`${id}-character-count`]: !!characterLimit,
            [`${id}-hint`]: !!hint,
          }) || undefined
        }
      />
      {characterLimit && (
        <CharacterCount
          characterLimit={characterLimit}
          id={`${id}-character-count`}
          numberOfCharacters={value.length}
        />
      )}
    </div>
  );
};
