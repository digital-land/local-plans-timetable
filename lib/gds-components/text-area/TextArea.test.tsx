import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { TextArea } from "./TextArea";

it("returns the plain text value", async () => {
  const onChange = jest.fn();
  const { asFragment } = render(
    <TextArea label="input" onChange={onChange} value="" />
  );

  await userEvent.type(screen.getByTestId("input-text-area"), "a");

  expect(asFragment()).toMatchSnapshot();
  expect(onChange).toHaveBeenCalledWith("a");
});
