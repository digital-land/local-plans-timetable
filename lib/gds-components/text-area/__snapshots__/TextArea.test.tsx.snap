// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the plain text value 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <label
      class="govuk-label"
    >
      input
    </label>
    <textarea
      class="govuk-textarea govuk-!-margin-bottom-2"
      data-testid="input-text-area"
      rows="5"
    />
  </div>
</DocumentFragment>
`;
